server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java

```diff
@@ -287,65 +291,8 @@ public void getPropertyInfo(
-            String businessTransaction,
-            @ApiParam(required = false,
-            value = "traces after this time,"
-                    + " millisecond since epoch") @DefaultValue("0") @QueryParam("startTime") long startTime,
-            @ApiParam(required = false,
-                    value = "traces before this time, "
-                            + "millisecond since epoch") @DefaultValue("0") @QueryParam("endTime") long endTime,
-            @ApiParam(required = false,
-                            value = "host name") @QueryParam("hostName") String hostName,
-            @ApiParam(required = false,
-                            value = "principal") @QueryParam("principal") String principal,
-            @ApiParam(required = false,
-                            value = "uri") @QueryParam("uri") String uri,
-            @ApiParam(required = false,
-                            value = "operation") @QueryParam("operation") String operation,
-            @ApiParam(required = false,
-                            value = "traces with these properties, defined as a comma "
-                                    + "separated list of name|value "
-                                    + "pairs") @DefaultValue("") @QueryParam("properties") String properties,
-            @ApiParam(required = false,
-                                    value = "faults") @QueryParam("faults") String faults) {
-
-        try {
-            Criteria criteria = new Criteria();
-            criteria.setBusinessTransaction(businessTransaction);
-            criteria.setStartTime(startTime);
-            criteria.setEndTime(endTime);
-            criteria.setHostName(hostName);
-            criteria.setPrincipal(principal);
-            criteria.setUri(uri);
-            criteria.setOperation(operation);
-
-            RESTServiceUtil.decodeProperties(criteria.getProperties(), properties);
-
-            RESTServiceUtil.decodeFaults(criteria.getFaults(), faults);
-
-            log.tracef("Get principal info for criteria [%s]", criteria);
-
-            java.util.List<PrincipalInfo> pis = analyticsService.getPrincipalInfo(
-                    securityProvider.validate(tenantId, context.getUserPrincipal().getName()), criteria);
-
-            log.tracef("Got principal info for criteria [%s] = [%s]", criteria, pis);
-
-            response.resume(Response.status(Response.Status.OK).entity(pis).type(APPLICATION_JSON_TYPE)
-                    .build());
-
-        } catch (Throwable e) {
-            log.debug(e.getMessage(), e);
-            Map<String, String> errors = new HashMap<String, String>();
-            errors.put("errorMsg", "Internal Error: " + e.getMessage());
-            response.resume(Response.status(Response.Status.INTERNAL_SERVER_ERROR)
-                    .entity(errors).type(APPLICATION_JSON_TYPE).build());
-        }
-
+    public void getPrincipalInfo(@Suspended final AsyncResponse response, @BeanParam PrincipalsGetRequest request) {
```

Comment:
Note that I replaced all the individual parameters by a `BeanParam`. While this requires an extra class, I believe this provides a good level of organization for the code. Besides, it allows for a generalization, useful for having the same actual handler code for both HTTP methods.